nixos/dovecot: More sieve options and precompiling of scripts

griff opened the pull request:
##### Motivation for this change

This PR does 4 different things in the 3 commit:

* Adds test of sieve scripts to `nixos/tests/dovecot.nix`
* Patches dovecot_pigeonhole with this change dovecot/pigeonhole#4
* Adds builder to precompile sieve scripts
* Adds more options relating to sieve scripts to dovecot2 module

###### test of sieve scripts

Before making changes to how sieve scripts were handled I thought it best that the dovecot test included some testing of the sieve script functionality.

######  Patching dovecot_pigeonhole and builder to precompile scripts

One of the changes I wanted to make was to precompile sieve scripts as part of a Nix derivation. But to do that dovecot_pigeonhole needs to be changed since when source and compiled script have same mtime it thinks that the compiled script is out of date. I have tried to upstream that change with dovecot/pigeonhole#4 but it is as of yet not merged.

The builder does its compilation in the install phase because the compiled `.svbin` files contain a reference to the full path of the source `.sieve` file and so the source needs to be in `$out` when doing the compile.

######  Sieve related options in the dovecot2 module

When precompiling sieve scripts the compiler needs a dovecot configuration that defines `sieve_plugins`, `sieve_extensions` and `sieve_global_extensions` to know which extensions and plugin functionality to allow but you can’t just use the dovecot2 config file directly because that would create a dependency cycle. So to better support precompilation I have added explicit options for these 3 things to the dovecot2 module so that I can reuse their values when precompiling.

The `sieveScripts` option followed very closely how dovecot configures global scripts and was a direct mapping from attribute name to `sieve_${attribute name}` which has some problems. E.g. if you defined `sieveScripts.after2` without defining `sieveScripts.after` your script would never be called. It also wasn’t totally clear without reading the dovecot documentation what scripts you could define.

So I have added explicit options `sieve.beforeScripts`, `sieve.defaultScript`, `sieve.afterScripts` and `sieve.discardScript` for each of the scripts you can define with `beforeScripts` and `afterScripts` being lists of paths that are automatically translated to the correct numbered dovecot configuration and with the added bonus that multiple definitions of the option are merged into one list. I have also included functionality so that existing configurations that use the `sieveScripts` option continue to work as before but now with a warning pointing to the new options.

The other place were you might use global sieve scripts and so would like to precompile the scripts is when using the `imapsieve` plugin. Before there were no options for this so I have added options for enabling and configuring `imapsieve` with support for automatic precompilation of scripts.


##### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

griff committed:
nixos/tests/dovecot: test sieveScripts option

griff committed:
dovecot_pigeonhole: patch and builder for precompiled sieve scripts

dotlambda commented:
/cc @andir @fpletz

uwap commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -131,6 +247,60 @@ let
+        '';
+      };
+      beforeScript = mkOption {
+        type = types.nullOr types.path;
```
What is the reason to ask for a path to a sieve script instead of an actual string containing the sieve script here?

wmertens replied:
> On the other hand if this option was a string and I prefer to store my sieve scripts in files I would have to use fileContents to load the data from the file, only for the module to write that content to a different file which seems a waste

Yes, but using a string is standard… and it's not like there's not a lot of copying going on during builds ;)

griff replied:
@wmertens I know. I guess I just like `writeText` more than `fileContents`.

I have considered supporting both `strings` and `path` by making the value an `attrset` like `environment.etc` or `services.rspamd.locals` but a `listOf attrset` is not as nice to work with as the named submodules like `environment.etc` and `services.rspamd.locals`.

wmertens replied:
How about having both text and path attributes? Or, if a Sieve script can never be a single line starting with `/`, autodetect?

griff replied:
@wmertens I looked at the Sieve RFC and from that it seems that while a sieve script can start with `/` it has to start with `/*` in that case which is obviously not a valid store path. So I am working on an update for this PR the resolves merge conflicts and supports having both text and path attributes for Sieve scripts.

wmertens commented:
I don't use Sieve but LGTM. I'm not that bothered by the paths vs text issue.

griff committed:
nixos/dovecot: more sieve options and precompile sieve scripts

griff commented:
I have now resolved all the merge conflicts and removed the commit that was no longer relevant since the functionality had since been implemented in another PR.

I have also done as @wmertens sugested and implemented support for Sieve scripts being either a string with the script or a path and the module will automatically detect the string script and write it to the store.

@uwap I hope this addresses the reservations you had.

nlewo commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -7,7 +7,90 @@ let
+        ${concatStringsSep "\n" (imap1 (i: v: ''
+          mkdir -p after/${toString i}
+          if [ -d '${v}' ]; then
+            cp -p '${v}'/*.sieve after/${toString i}
```
Would it be possible to create symlinks instead of copying files?

griff replied:
Done! And while looking at this I found that when the path was outside the store the user feedback wasn't that great and would result in an unclean build in non-sandboxed builds so I have added logic to use `fileContents` to write paths outside the store to the store first.

nlewo commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -7,7 +7,90 @@ let
+      plugins = cfg.sieve.plugins;
+      extensions = cfg.sieve.extensions;
+      globalExtensions = cfg.sieve.globalExtensions;
+      phases = ["buildPhase" "installPhase" "fixupPhase"];
```
It is recommended to use `dontConfigure` instead of defining your own list of phases.

griff replied:
done! But needed to also set `dontUnpack`

nlewo commented on pkgs/servers/mail/dovecot/plugins/pigeonhole/default.nix:
```diff
@@ -11,6 +11,15 @@ stdenv.mkDerivation rec {
   buildInputs = [ dovecot openssl ];
 
+  patches = [
+    # Fix mtime comparison so that scripts can be precompiled
```
Could you link your PR also?

griff replied:
done!

nlewo commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -7,7 +7,90 @@ let
-  stateDir = "/var/lib/dovecot";
+
+  sieveToPath = text:
+    if isString text && builtins.substring 0 2 text == "/*"
```
The  `lib.hasPrefix` would be more explicit.

griff replied:
done.

nlewo reviewed (commented):
These changes are really nice! 
Here some minor comments and i still would like to play with this PR a bit to see what is generated and better understand it.

SuperSandro2000 commented on pkgs/servers/mail/dovecot/plugins/pigeonhole/compile.nix:
```diff
@@ -0,0 +1,38 @@
+      else
+        echo "Compiling $k -> $(readlink -f "$k")"
+      fi
+      ${dovecot_pigeonhole}/bin/sievec -c "${config}" "$out/sieve/$k"
```
```suggestion
     sievec -c "${config}" "$out/sieve/$k"
```

This should use nativeBuildInputs where we add  dovecot_pigeonhole.

griff replied:
Done!

RaitoBezarius commented:
Can you fix the conflicts please? I think this PR is ready to merge.

griff commented:
@RaitoBezarius this has been ready to merge for years so I wasn't that motivated to rebase it again and again. In the meantime the patch to pigeonhole has been rejected so I also needed to move the 2 line patch into nixpkgs.

RaitoBezarius commented:
Apologies for the delay, I missed your message… @2xsaiko became maintainer of that module, can I let you shepherd this PR and give me the go for a merge?

2xsaiko commented:
Sure, I'll take a look in a bit!

2xsaiko commented on pkgs/servers/mail/dovecot/plugins/pigeonhole/compile.nix:
```diff
@@ -0,0 +1,38 @@
+  nativeBuildInputs = nativeBuildInputs ++ [ dovecot_pigeonhole ];
+  installPhase = ''
+    mkdir -p "$out/sieve/"
+    for k in $(find * -name \*.sieve); do
+      mkdir -p "$out/sieve/$(dirname "$k")"
+      cp -a "$k" "$out/sieve/$k"
+      if [[ ! -L "$k" ]]; then
+        echo "Compiling $k"
+        touch -m --date=@0 "$out/sieve/$k"
+      else
+        echo "Compiling $k -> $(readlink -f "$k")"
+      fi
+      sievec -c "${config}" "$out/sieve/$k"
+    done
```
```suggestion
    while IFS= read -r k; do
      mkdir -p "$out/sieve/$(dirname "$k")"
      cp -a "$k" "$out/sieve/$k"
      if [[ ! -L "$k" ]]; then
        echo "Compiling $k"
        touch -m --date=@0 "$out/sieve/$k"
      else
        echo "Compiling $k -> $(readlink -f "$k")"
      fi
      sievec -c "${config}" "$out/sieve/$k"
    done < <(find -name '*.sieve')
```

to handle files with spaces

2xsaiko commented on nixos/modules/services/mail/dovecot.nix:
```diff
@@ -7,7 +7,94 @@ let
+    else pkgs.writeText "script.sieve" text;
+  scriptType = types.coercedTo types.lines sieveToPath types.path;
+
+  compileSieveScripts = pkgs.buildSieveScripts {
```
```suggestion
  compiledSieveScripts = pkgs.buildSieveScripts {
```

Does compiling these in advance break `include` for non-store paths? Someone may be using that for dynamically generated rules. If it does, would they be doable another way?

dblsaiko reviewed (changes requested):
Please rebase.

dotlambda commented:
Can this be closed now that https://github.com/NixOS/nixpkgs/pull/275031 was merged?
